### ISSUE TITLE
no-dupe-else-if inappropriately triggering if two conditions are identical in terms of code, but are not deterministic, so it is not true that the second branch will never run

### ISSUE THREAD

**Issue by dumbmatter** [eslint/eslint#13279]
**Tell us about your environment**

* **ESLint Version:** v7.0.0
* **Node Version:** v14.1.0
* **npm Version:** v6.14.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
    "parserOptions": {
        "ecmaVersion": 5,
        "sourceType": "script",
        "ecmaFeatures": {}
    },
    "rules": {
        "no-dupe-else-if": 2
    },
    "env": {
        "browser": true
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
if (Math.random() < 0.5) {
  console.log('hi');
} else if (Math.random() < 0.5) {
  console.log('hey');
}
```

https://eslint.org/demo#eyJ0ZXh0IjoiaWYgKE1hdGgucmFuZG9tKCkgPCAwLjUpIHtcbiAgY29uc29sZS5sb2coJ2hpJyk7XG59IGVsc2UgaWYgKE1hdGgucmFuZG9tKCkgPCAwLjUpIHtcbiAgY29uc29sZS5sb2coJ2hleScpO1xufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6NSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby1kdXBlLWVsc2UtaWYiOjJ9LCJlbnYiOnsiYnJvd3NlciI6dHJ1ZX19fQ==

**What did you expect to happen?**

There should be no error. Or if an error is desired because people deem that code to be bad, it should not say "this branch can never execute" because it clearly can execute.

**What actually happened? Please include the actual, raw output from ESLint.**

Error message:

> 3:12 - This branch can never execute. Its condition is a duplicate or covered by previous conditions in the if-else-if chain. (no-dupe-else-if)

**Are you willing to submit a pull request to fix this bug?**

No sorry, not familiar with the codebase and don't have the time to learn right now :)

**Comment by anikethsaha** [comment]
Thanks for the issue, 

IMO, eslint doesnt know about whats the operation being done by the function/method. so even if we can have a list of known function which we want to ignore like for `Math.random` as you mentioned, there might be some user-defined function like this

```js
const foo = () =>  Math.random()
if (foo() < 0.5) {
  console.log('hi');
} else if (foo() < 0.5) {
  console.log('hey');
}
```

it would still fail even if we have `Math.random` in ignore list or something like this

```js
if(callToApiReturnRandomData() === 'a'){

}else if (callToApiReturnRandomData() === 'a'){

}
```

it would still fail, 

so I would suggest having some truth condition with `||` or some non-changing integer/char addition which wont effect the performance at the same time it will be lint free.

like this 

```js
if (Math.random() < 0.5) {
  console.log('hi');
} else if (Math.random() < 0.5 || true) { // lint free
  console.log('hey');
}
```
[repl](https://eslint.org/demo#eyJ0ZXh0IjoiaWYgKE1hdGgucmFuZG9tKCkgPCAwLjUpIHtcbiAgY29uc29sZS5sb2coJ2hpJyk7XG59IGVsc2UgaWYgKE1hdGgucmFuZG9tKCkgPCAwLjUgfHwgdHJ1ZSkge1xuICBjb25zb2xlLmxvZygnaGV5Jyk7XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo1LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLWR1cGUtZWxzZS1pZiI6Mn0sImVudiI6eyJicm93c2VyIjp0cnVlfX19)

or like this 

```js
if (Math.random() < 0.5) {
  console.log('hi');
} else if (Math.random() < (0.5 * 1)) { // lint free
  console.log('hey');
}
```

[repl](https://eslint.org/demo#eyJ0ZXh0IjoiaWYgKE1hdGgucmFuZG9tKCkgPCAwLjUpIHtcbiAgY29uc29sZS5sb2coJ2hpJyk7XG59IGVsc2UgaWYgKE1hdGgucmFuZG9tKCkgPCAoMC41ICogMSkpIHtcbiAgY29uc29sZS5sb2coJ2hleScpO1xufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6NSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby1kdXBlLWVsc2UtaWYiOjJ9LCJlbnYiOnsiYnJvd3NlciI6dHJ1ZX19fQ==)

**Comment by mdjermanovic** [comment]
@dumbmatter thanks for the bug report!

I think this is a good case for an `eslint-disable-*` comment, as described in the [When Not To Use It](https://eslint.org/docs/rules/no-dupe-else-if#when-not-to-use-it) section.

As for the message, I don't have a strong opinion. It's indeed incorrect in this particular case, but it usually provides useful additional info about the effect of the error in the code, since the rule doesn't report only identical conditions:

```js
if (a || b) {
    foo();
} else if (a) { // error
    bar();
}
```
